src/lua/funcs/mod.rs

```diff
@@ -8,14 +8,29 @@ use hlua::{Lua, LuaTable, Function};
 #[forbid(unused_variables)]
 
 mod input;
+mod lua_registry;
 
 /// Register all the Rust functions for the lua libraries
 pub fn register_libraries(lua: &mut Lua) {
     // Yeah, need to access individual tables
 
-    let mut wm: LuaTable<_> = lua.get("wm").unwrap();
-    let mut pointer: LuaTable<_> = wm.get("pointer").unwrap();
+    {
+        let mut wm: LuaTable<_> = lua.get("wm").unwrap();
+        let mut pointer: LuaTable<_> = wm.get("pointer").unwrap();
 
-    pointer.set("get_position", hlua::function0(input::pointer_get_position));
-    pointer.set("set_position", hlua::function2(input::pointer_set_position));
+        pointer.set("get_position", hlua::function0(input::pointer_get_position));
+        pointer.set("set_position", hlua::function2(input::pointer_set_position));
+    }
+
+    // Initialize registry at the end?
+    init_registry(lua);
+}
+
+#[inline] // It's called once
+fn init_registry(lua: &mut Lua) {
+    let mut reg_table: LuaTable<_> = lua.get("registry").unwrap();
+    let mut meta_reg = reg_table.get_or_create_metatable();
+    meta_reg.set("__metatable", "Turtles all the way down");
+    meta_reg.set("__index", hlua::function2(lua_registry::index));
+    meta_reg.set("__newindex", hlua::function3(lua_registry::new_index));
```

Comment:
We are overloading three metatable methods: `__index`, `__index_new`, and `__metatable`.

So a metatable (for the purposes of the registry table) is basically a list of operators. For example, to make the registry look like a Lua table, we use the metatable to implement it's "operator\_[]\_get" ( `x = registry[key]` uses the `index` method) and set "operator_[]_set" (`registry[key] = val` uses the `index_new` method).

The `__metatable` method is a way to prevent users from changing the `registry` to become an actual table using a new metatable.

Metatables also let you introduce OOD patterns and things like polymorphism - you can use the metatable to say things like "every time an access is made to a table, try to access the field on this other table first" - a la superclasses.

(This is similar to Javascript objects having an `__proto__` field, except it's designed in a way that you cannot access a table's metatable from the table itself).

The registry may be the only structure with a metatable, if you need other Rust data you'll probably use a function.